packages/ui/src/app/pages/Election/components/ElectionProgressBar.tsx

```diff
@@ -0,0 +1,459 @@
+    revealingDays = blockDurationToDays(constants?.election.revealingPeriod)
+    inactiveDays = blockDurationToDays(constants?.idlePeriod)
+
+    // calculate progress bar layout
+    progressBarAttr = `${inactiveDays > 0 ? inactiveDays : 1}fr ${announcingDays > 0 ? announcingDays : 1}fr ${
+      votingDays > 0 ? votingDays : 1
+    }fr ${revealingDays > 0 ? revealingDays : 1}fr`
+
+    // calculate progress status variables
+    const stages = ['inactive', 'announcing', 'voting', 'revealing']
+    const currentPosition = stages.indexOf(props.electionStage)
+    const gaps = [
+      constants?.announcingPeriod,
+      constants?.election.votingPeriod,
+      constants?.election.revealingPeriod,
+      A_DAY / MILLISECONDS_PER_BLOCK - 1,
+    ]
+
+    inactiveProgress = [Math.floor(100 - (100 * duration) / constants?.idlePeriod), 100, 100, 100][currentPosition]
+    announcingProgress = [0, Math.floor(100 - (100 * duration) / constants?.announcingPeriod), 100, 100][
+      currentPosition
+    ]
+    votingProgress = [0, 0, Math.floor(100 - (100 * duration) / constants?.election.votingPeriod), 100][currentPosition]
+    revealingProgress = [0, 0, 0, Math.floor(100 - (100 * duration) / constants?.election.revealingPeriod)][
+      currentPosition
+    ]
+    remainDays = blockDurationToDays(duration + sum(gaps.slice(currentPosition)))
+
+    inactiveEndDay = blockToDate(
+      [duration, duration - sum(gaps.slice(0, 1)), duration - sum(gaps.slice(0, 2)), duration - sum(gaps.slice(0, 3))][
+        currentPosition
+      ]
+    )
+
+    announcingEndDay = blockToDate(
+      [duration + sum(gaps.slice(0, 1)), duration, duration - sum(gaps.slice(1, 2)), duration - sum(gaps.slice(1, 3))][
+        currentPosition
+      ]
+    )
+
+    votingEndDay = blockToDate(
+      [duration + sum(gaps.slice(0, 2)), duration + sum(gaps.slice(1, 2)), duration, duration - sum(gaps.slice(2, 3))][
+        currentPosition
+      ]
+    )
+
+    revealingEndDay = blockToDate(
+      [duration + sum(gaps.slice(0, 3)), duration + sum(gaps.slice(1, 3)), duration + sum(gaps.slice(2, 3)), duration][
+        currentPosition
+      ]
+    )
+
+    inactiveEndBlock =
+      currentBlock +
+      [duration, duration - sum(gaps.slice(0, 1)), duration - sum(gaps.slice(0, 2)), duration - sum(gaps.slice(0, 3))][
+        currentPosition
+      ]
+
+    announcingEndBlock =
+      currentBlock +
+      [duration + sum(gaps.slice(0, 1)), duration, duration - sum(gaps.slice(1, 2)), duration - sum(gaps.slice(1, 3))][
+        currentPosition
+      ]
+
+    votingEndBlock =
+      currentBlock +
+      [duration + sum(gaps.slice(0, 2)), duration + sum(gaps.slice(1, 2)), duration, duration - sum(gaps.slice(2, 3))][
+        currentPosition
+      ]
+
+    revealingEndBlock =
+      currentBlock +
+      [duration + sum(gaps.slice(0, 3)), duration + sum(gaps.slice(1, 3)), duration + sum(gaps.slice(2, 3)), duration][
+        currentPosition
+      ]
```

Comment:
Hi @mkbeefcake the change I suggested here: https://github.com/Joystream/pioneer/pull/4556#discussion_r1409691009 is not about `useCouncilRemainingPeriods`. It's fine to have the `useCouncilPeriodInformation` hook (maybe it should be in it's own file but that's a detail). What I suggested was to change the `useCouncilPeriodInformation` some more so it can simplify this code.

The goal of this suggestion is to change this part of the code to something like this:

```suggestion
    const [inactiveEndBlock, announcingEndBlock, votingEndBlock, revealingEndBlock] =
      periodInformation?.periodEnds ?? []

    const endDates = periodInformation?.periodEnds.map((block) => blockToDate(block, currentBlock))
    const [inactiveEndDay, announcingEndDay, votingEndDay, revealingEndBlock] =
      endDates ?? []
      
    const progresses = periodInformation && periodInformation.periodEnds.map((end, index) => {
      const start = periodInformation.periodStarts[index]
      return clamp((currentBlock - start) * 100 / (end - start), 0, 100)
    })
    const [inactiveProgress, announcingProgress, votingProgress, revealingProgress] =
      progresses ?? []

    const remainDays = periodInformation && blockDurationToDays(periodInformation.remainingPeriod)
```
Here I put `const ...` because this could be moved out of the `if` (so there would be no need for all the `let ...`).

Also `periodInformation` is the output of the change I suggested for `useCouncilPeriodInformation`. I think this hook can be called in this component directly (no need to call it in `packages/ui/src/app/pages/Election/Election.tsx`).

Finally `blockToDate` is the version I suggested which converts a block height into a date (and not a block duration to a date), and [`clamp`](https://lodash.com/docs/4.17.15#clamp) from Lodash.